### PR TITLE
Preserve EJson type metadata in SSTs

### LINKED ISSUE #2354: SST timestamp values are turned into maps
From some test data like this:

``` json
{ "testTimestamp": { "$timestamp": "2004-10-19T10:23:54Z" } }
{ "testTimestamp": { "$timestamp": "2014-10-19T10:23:54Z" } }
```

The returned structure is:

``` json
            "testTimestamp": {
                "measure": {
                    "count": 2,
                    "minLength": 7,
                    "maxLength": 7
                },
                "structure": {
                    "type": "map",
                    "of": {
                        "minute": {
                            "measure": {
                                "count": 2,
                                "distribution": {
                                    "mean": 23,
                                    "variance": 0
                                },
                                "min": 23,
                                "max": 23
                            },
                            "structure": {
                                "type": "const",
                                "of": 23
                            }
                        },
                        "year": {
                            "measure": {
                                "count": 2,
                                "distribution": {
                                    "mean": 2009,
                                    "variance": 50
                                },
                                "min": 2004,
                                "max": 2014
                            },
                            "structure": {
                                "type": "integer"
                            }
                        },
                        "hour": {
                            "measure": {
                                "count": 2,
                                "distribution": {
                                    "mean": 10,
                                    "variance": 0
                                },
                                "min": 10,
                                "max": 10
                            },
                            "structure": {
                                "type": "const",
                                "of": 10
                            }
                        },
                        "second": {
                            "measure": {
                                "count": 2,
                                "distribution": {
                                    "mean": 54,
                                    "variance": 0
                                },
                                "min": 54,
                                "max": 54
                            },
                            "structure": {
                                "type": "const",
                                "of": 54
                            }
                        },
                        "day_of_month": {
                            "measure": {
                                "count": 2,
                                "distribution": {
                                    "mean": 19,
                                    "variance": 0
                                },
                                "min": 19,
                                "max": 19
                            },
                            "structure": {
                                "type": "const",
                                "of": 19
                            }
                        },
                        "nanosecond": {
                            "measure": {
                                "count": 2,
                                "distribution": {
                                    "mean": 0,
                                    "variance": 0
                                },
                                "min": 0,
                                "max": 0
                            },
                            "structure": {
                                "type": "const",
                                "of": 0
                            }
                        },
                        "month": {
                            "measure": {
                                "count": 2,
                                "distribution": {
                                    "mean": 10,
                                    "variance": 0
                                },
                                "min": 10,
                                "max": 10
                            },
                            "structure": {
                                "type": "const",
                                "of": 10
                            }
                        }
                    }
                }
            }
```

It seems that the timestamp is treated as a map of date/time components - which is fine, and could well be useful, but the information that it was ever a timestamp has vanished.

I'm not sure what happens with plain date types, as they're unrepresentable in mongo, and get turned into a timestamp with a zeroed-out time component.

### LINKED ISSUE #2356: SST interval values are turned into decimals
From some test data like this:

``` json
{ "testInterval": { "$interval": "P7DT12H34M" } }
{ "testInterval": { "$interval": "P3DT12H34M" } }
```

The returned structure is:

``` json
            "testInterval": {
                "measure": {
                    "count": 2,
                    "distribution": {
                        "mean": 477240000,
                        "variance": 59719680000000000
                    },
                    "min": 304440000,
                    "max": 650040000
                },
                "structure": {
                    "type": "decimal"
                }
            }
```

### LINKED ISSUE #2357: SST ObjectId values are turned into strings
From some test data like this:

``` json
{ "testObjectId": { "$oid": "507f1f77bcf86cd799439011" } }
{ "testObjectId": { "$oid": "7f1f77bcf86cd79943901150" } }
```

The returned structure is:

``` json
            "testObjectId": {
                "measure": {
                    "count": 2,
                    "min": "507f1f77bcf86cd799439011",
                    "max": "7f1f77bcf86cd79943901150",
                    "minLength": 24,
                    "maxLength": 24
                },
                "structure": {
                    "type": "array",
                    "of": {
                        "measure": {
                            "count": 2
                        },
                        "structure": {
                            "type": "character"
                        }
                    }
                }
            }
```

### LINKED ISSUE #2358: SST json encoding - precise?
This one is more of a question... and at the moment a redundant one at that! After #2354 #2355 #2356 #2357 it will be important though.

Because of the above issues I've not been able to have the `/schema` endpoint produce a value that would need `precise` encoding for it to be unambiguous, but I just wanted to double check whether that is the case - the `Accept` header seems to be ignored on this endpoint.

### LINKED ISSUE #2363: Preserve type metadata in SSTs
Generating SSTs should take type metadata (`{"_ejson.type" : ...}`) into account and only merge structure where the type is equal, when present, preserving the tagged structure as unique otherwise.

Motivated by #2354, #2355, #2356 and #2357.

### PR DESCRIPTION
**Pull request by wemrysi**
Updates the representation of SSTs to allow for preserving `_ejson.type` metadata from the source `EJson`.

Closes #2354.
Closes #2356.
Closes #2357.
Closes #2358.
Closes #2363.

### CODE REVIEW

**Comment by wemrysi** [comment]
@drostron All the commits should be mostly self-contained and make sense in isolation. In hindsight, this probably should have been 4 PRs, one for the preparatory SST changes that don't specifically relate to #2363, one organizing the EJson type tags, one updating our testing framework for recursive structures and then this PR. I'll try and be more disciplined about breaking things apart in the future.

File: frontend/src/main/scala/quasar/sst/compression.scala [hunk 1]
```diff
@@ -242,9 +213,18 @@ object compression {
   )(implicit
     JC: Corecursive.Aux[J, EJson],
     JR: Recursive.Aux[J, EJson]
-  ): SST[J, A] =
-    sstConst[J, A].headOption(sst).fold(sst) { j =>
-      val ts = sstMeasure[J, A].get(sst)
-      sstMeasure[J, A].set(ts)(primarySST(ts.size, j))
+  ): SST[J, A] = {
+    def psst(ts: TypeStat[A], j: J): SST[J, A] =
```
Comment (drostron) [inline frontend/src/main/scala/quasar/sst/compression.scala]: 👂

Comment (wemrysi) [inline frontend/src/main/scala/quasar/sst/compression.scala]: LOL

File: foundation/src/test/scala/quasar/contrib/matryoshka/UWidth.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * limitations under the License.
+ */
+
+package quasar.contrib.matryoshka
+
+import slamdata.Predef._
+
+import scalaz._
+
+sealed abstract class UWidth[F[_]] {
```
Comment (wemrysi) [inline foundation/src/test/scala/quasar/contrib/matryoshka/UWidth.scala]: Union. Apparently I was in an economical mood, I'll rename to `UnionWidth`.

File: frontend/src/main/scala/quasar/sst/Tagged.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import slamdata.Predef._
+
+import matryoshka._
+import scalaz._, Scalaz._
+
+// TODO: More appropriate name than `Tagged`?
+final case class Tagged[A](tag: String, value: A)
+
+object Tagged {
+  implicit val traverse1: Traverse1[Tagged] =
```
Comment (wemrysi) [inline frontend/src/main/scala/quasar/sst/Tagged.scala]: Yes indeed, will add them.

**Comment by drostron** [comment]
@wemrysi Needs a version label

**Pull request merged**